src/safir/slack.py

```diff
@@ -0,0 +1,488 @@
+from structlog.stdlib import BoundLogger
+
+from safir.datetime import current_datetime, format_datetime_for_logging
+from safir.dependencies.http_client import http_client_dependency
+
+__all__ = [
+    "SlackClient",
+    "SlackException",
+    "SlackIgnoredException",
+    "SlackField",
+    "SlackIgnoredException",
+    "SlackMessage",
+    "SlackRouteErrorHandler",
+]
+
+
+class SlackField(BaseModel):
+    """A component of a Slack message with a heading.
+
+    Intended for use in the ``fields`` portion of a Block Kit message. If the
+    formatted output is longer than 2000 characters, it will be truncated to
+    avoid the strict upper limit imposed by Slack.
+    """
+
+    heading: str
+    """Heading of the field (shown in bold)."""
+
+    text: Optional[str] = None
+    """Text of the field as normal text (use this or ``code``).
+
+    This is always marked as vertabim, so channel mentions or @-mentions of
+    users will not be treated as special.
+    """
+
+    code: Optional[str] = None
+    """Text of the field as a code block (use this or ``text``)."""
+
+    max_formatted_length: ClassVar[int] = 2000
+    """Maximum length of formatted output, imposed by Slack.
+
+    Used internally for validation and intended to be overridden by child
+    classes that need to impose different maximum lengths.
+    """
+
+    @root_validator
+    def _validate_content(cls, values: Dict[str, Any]) -> Dict[str, Any]:
+        """Ensure only one of ``text`` or ``code`` is set."""
+        if values["text"] and values["code"]:
+            raise ValueError("Only one of text and code may be given")
+        if values["text"] is None and values["code"] is None:
+            raise ValueError("One of text or code must be given")
+        return values
+
+    def to_slack(self) -> Dict[str, Any]:
+        """Convert to a Slack Block Kit block.
+
+        Returns
+        -------
+        dict
+            A Slack Block Kit block suitable for including in the ``fields``
+            or ``text`` section of a ``blocks`` element.
+        """
+        heading = f"*{self.heading}*\n"
+        if self.code:
+            extra_needed = len(heading) + 8  # ```\n\n```
+            max_length = self.max_formatted_length - extra_needed
+            code = _format_and_truncate_at_start(self.code, max_length)
+            body = f"```\n{code}\n```"
+        else:
+            if not self.text:
+                # Pydantic validation should make this impossible.
+                raise RuntimeError("SlackField without code or text")
+            extra_needed = len(heading)
+            max_length = self.max_formatted_length - extra_needed
+            body = _format_and_truncate_at_end(self.text, max_length)
+        return {"type": "mrkdwn", "text": heading + body, "verbatim": True}
+
+
+class SlackAttachment(SlackField):
+    """An attachment in a Slack message with a heading.
+
+    Intended for use in the ``attachments`` portion of a Block Kit message. If
```

Comment:
Do we want to mention anywhere that attachments are considered legacy by Slack, so they might be made less visible over time?